### PR TITLE
Change override to override_sole_init in alf.config

### PR DESCRIPTION
**Pull request by QuantuMope**
This PR changes the `override` flag of `alf.config` to `override_sole_init`. The implementation before allowed `alf.override_config` to override a value of a config regardless of the `mutable` flag. Unfortunately, this breaks the logic of setting parameters via `--conf_params` or `alf.pre_config`. It also causes issues in evaluator processes that may set certain config values with `mutable=False` and then read from the conf file as it overwrites the desired eval setting.

Therefore, this PR now changes `alf.override_config` to `alf.override_sole_config`, which respects immutability. On our local hobot code, I've now verified that all the above problems have been remedied and singular config calls can now be enforced when using `ALF_SOLE_CONFIG=1`, greatly reducing chance for bugs via nested / unintentional overwrites.

Additional test cases for edge cases has also been provided in `config_utils_test.py`.

Once this gets merged, I can submit PRs for `ALF_SOLE_CONFIG=1` compatible hobot confs.

### CODE REVIEW

File: alf/config_util_test.py [hunk 1]
```diff
@@ -182,6 +182,63 @@ def sole_init_test_env(x):
         self.assertEqual(alf.get_config_value("sole_init_test_twice.x"), 1)
         self.assertEqual(alf.get_config_value("sole_init_test_env.x"), 1)
 
+        # Test override_config doesn't doesn't overwrite for immutable values.
```
Comment (le-horizon) [inline alf/config_util_test.py]: remove one "doesn't"

File: alf/config_util_test.py [hunk 2]
```diff
@@ -182,6 +182,63 @@ def sole_init_test_env(x):
...
+        @alf.configurable
+        def override_on_immutable(x):
+            pass
+
+        alf.config("override_on_immutable", x=0, mutable=False)
+        alf.override_config("override_on_immutable", x=1)
```
Comment (le-horizon) [inline alf/config_util_test.py]: Is it possible to raise an error on overriding immutable config?  Silently ignoring override can be hard to debug?

File: alf/config_util_test.py [hunk 3]
```diff
@@ -182,6 +182,63 @@ def sole_init_test_env(x):
...
+            alf.get_config_value("override_on_immutable_and_sole_init.x"), 0)
+
+        # Test that pre_config calls work before a sole_init config call
+        @alf.configurable
+        def pre_config_before(x):
+            pass
+
+        alf.pre_config({"pre_config_before.x": 0})
+        alf.config("pre_config_before", x=1)
+        alf.override_config("pre_config_before", x=1)
```
Comment (le-horizon) [inline alf/config_util_test.py]: We use pre_config in deployment app to specify deploy mode etc..  In order to avoid them from being overridden, should we specify them as immutable?  Or should this override_config call throw error?

File: alf/config_util_test.py [hunk 4]
```diff
@@ -182,6 +182,63 @@ def sole_init_test_env(x):
...
+        with self.assertRaises(RuntimeError) as context:
+            alf.pre_config({"pre_config_after.x": 0})
+
+        # Test that calling override_config doesn't affect previous sole_init calls.
+        @alf.configurable
+        def override_no_affect_sole_init(x):
+            pass
+
+        alf.config("override_no_affect_sole_init", x=1, sole_init=True)
+        alf.override_config("override_no_affect_sole_init", x=2)
```
Comment (le-horizon) [inline alf/config_util_test.py]: should this also throw error?

File: alf/config_util.py [hunk 5]
```diff
@@ -60,7 +60,7 @@ def config(prefix_or_dict,
            mutable=True,
            raise_if_used=True,
            sole_init=False,
-           override=False,
+           override_sole_init=False,
```
Comment (le-horizon) [inline alf/config_util.py:63]: please also add tests for override_sole_init.

File: alf/config_util.py [hunk 6]
```diff
@@ -112,19 +112,18 @@ def func(self, a, b):
...
+            This is helpful in enforcing a singular point of initialization,
+            thus eliminating any potential side effects from possible future
+            overrides. For users wanting this to be the default behavior, the
+            ALF_SOLE_CONFIG env variable can be set to 1.
+        override_sole_init (bool): If True, the value of the config will be set
+            regardless of any previous ``sole_init`` setting. This should be used
+            only when absolutely necessary (e.g., a teacher-student training loop,
+            where the student must override certain configs inherited from the
+            teacher). If the config is immutable, a warning will be declared with
+            no changes made.
```
Comment (le-horizon) [inline alf/config_util.py:126]: instead of quietly ignoring the config, should we just throw error for this case?

**Review by le-horizon — COMMENTED** [review]
Thanks for the much cleaner changes and tests, Andrew.  A few questions regarding alf config behavior.

Le

File: alf/config_util.py (hunk 5 above)
Comment (QuantuMope) [reply alf/config_util.py:63]: `alf.override_config` is simply a wrapper for `alf.config(override_sole_init=True)`.

File: alf/config_util.py (hunk 6 above)
Comment (QuantuMope) [reply alf/config_util.py:126]: So the reason we cannot do this is because `alf.pre_config` and `--conf_param` set values with `mutable=False` (that is how those values avoid getting overwritten). Therefore, there may be times we want to override our conf files using one of the two options and raising an error in this situation is problematic. A workaround for this was to possibly introduce a `pre_configged` state variable for configs, but I thought that may be unnecessary. Let me know what you think.

File: alf/config_util_test.py (hunk 1 above)
Comment (QuantuMope) [reply alf/config_util_test.py]: Thanks! Done.

File: alf/config_util_test.py (hunk 2 above)
Comment (QuantuMope) [reply alf/config_util_test.py]: Refer to my reply above for why this cannot be done currently without adding more state parameters. There is a logger warning that gets printed in this scenario, though it's easy for users to not see it.

File: alf/config_util_test.py (hunk 3 above)
Comment (QuantuMope) [reply alf/config_util_test.py]: `pre_config` already sets all values as `mutable=False`. This is how it and `--conf_params` prevent their config values from being overwritten.

File: alf/config_util_test.py (hunk 4 above)
Comment (QuantuMope) [reply alf/config_util_test.py]: It shouldn't throw an error since `override_config` allows us to set a config's value despite an earlier `sole_init=True` call.

**Review by QuantuMope — COMMENTED** [review]
Thanks for the really speedy review Le. Addressed your comments. Let me know if there are any questions or clarifications I can make.

File: alf/config_util.py (hunk 6 above)
Comment (le-horizon) [reply alf/config_util.py:126]: I see.  You are right.  Let's go with current way for the PR.

Separately, are there some examples of these cases?  It would be good to understand a bit more, and see if we can design better solutions.

File: alf/config_util_test.py (hunk 2 above)
Comment (le-horizon) [reply alf/config_util_test.py]: Maybe let's color code the warning message a bright yellow, so it's harder to miss?

File: alf/config_util_test.py (hunk 3 above)
Comment (le-horizon) [reply alf/config_util_test.py]: Ah, my mistake.

File: alf/config_util_test.py [hunk 7]
```diff
@@ -182,6 +182,63 @@ def sole_init_test_env(x):
...
+        with self.assertRaises(RuntimeError) as context:
+            alf.config("override_no_affect_sole_init", x=3)
```
Comment (le-horizon) [inline alf/config_util_test.py:240]: test that x's value is 2

File: alf/config_util_test.py (hunk 4 above)
Comment (le-horizon) [reply alf/config_util_test.py]: Oh, I see.  Maybe rename the function to alf.override_sole_config()?  Just as how you renamed override to override_sole_init.

**Review by le-horizon — COMMENTED** [review]
Thanks for answering the questions.  Only a few nits.

Thanks,
Le

File: alf/config_helpers.py [hunk 8]
```diff
@@ -285,7 +285,11 @@ def get_env():
             # A 'None' random seed won't set a deterministic torch behavior.
             for _ in range(PerProcessContext().ddp_rank):
                 random_seed = random.randint(0, 2**32)
-        config1("TrainerConfig.random_seed", random_seed, raise_if_used=False)
+        config1(
```
Comment (emailweixu) [inline alf/config_helpers.py:296]: there are several other config1 calls in config_helpers.py that may need fix too.

Comment (QuantuMope) [reply alf/config_helpers.py:296]: Thanks for the catch. This has been fixed for the other calls.

File: alf/config_util.py (hunk 6 above)
Comment (QuantuMope) [reply alf/config_util.py:126]: Are you referring to examples of the teacher-student setup?

One is our own agent training. Let's say you train expert with `ALF_SOLE_CONFIG=1`. When we train the agent with `ALF_SOLE_CONFIG=1` as well, it will inevitably raise errors as agent often overwrites some of the expert's configs. Here, we can use the `alf.override_config` function. Not only does this minimize config calls, but it also explicitly tells the coder which inherited calls are being overwritten.

File: alf/config_util_test.py (hunk 2 above)
Comment (QuantuMope) [reply alf/config_util_test.py]: Right now it prints as red via `absl.logging.warning`. Do you think yellow is better?

File: alf/config_util_test.py (hunk 4 above)
Comment (QuantuMope) [reply alf/config_util_test.py]: Good idea. Done.

File: alf/config_util_test.py (hunk 7 above)
Comment (QuantuMope) [reply alf/config_util_test.py:240]: Thanks. Done.

**Review by QuantuMope — COMMENTED** [review]
Thanks for the comments. PTAL

**Review by emailweixu — APPROVED** [review]

**Review by le-horizon — APPROVED** [review]
